src/transformers/trainer.py

```diff
@@ -997,11 +1019,16 @@ def _save_checkpoint(self, model, trial, metrics=None):
             output_dir = os.path.join(self.args.output_dir, checkpoint_folder)
 
             self.store_flos()
+
         self.save_model(output_dir)
+        if self.deepspeed:
+            self.deepspeed.save_checkpoint(output_dir)
 
         # Save optimizer and scheduler
         if self.sharded_dpp:
             self.optimizer.consolidate_state_dict()
+        if self.deepspeed:
+            pass  # deepspeed.save_checkpoint above saves model/optim/sched
```

Comment:
What's the point of the test then (since there is no else?) We can put the comment at line 1024 above.

Reply:
the point is not to run the other branches - i.e. it's part of the logic. does it make sense?

Reply:
I don't think so, since it's really just after the previous `if self.deepspeed`. I think having the comment there and removing that test is clearer.

Reply:
I don't think so, since it's really just after the previous if self.deepspeed. I think having the comment there and removing that test is clearer.

Reply:
Please have a closer look. If I remove the branch, this branch will run:
```
        elif self.is_world_process_zero():
            torch.save(self.optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
            with warnings.catch_warnings(record=True) as caught_warnings:
                torch.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
            reissue_pt_warnings(caught_warnings)
```

Reply:
so then I need to stick an `and is not self.deepspeed` there.

I thought it was easier to read how I coded it, but either way works if you prefer the way above (the and ... code)

Reply:
The `elif` goes with `if is_torch_tpu_available():` as far as I can see, so I don't get your point.

Reply:
oh, oh! I see - the no line breaks confused me - I read the whole block as a single if/elif/else :( 

So 2 ways then:

1.
```
        if is_torch_tpu_available():
            xm.rendezvous("saving_optimizer_states")
            xm.save(self.optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
            with warnings.catch_warnings(record=True) as caught_warnings:
                xm.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
                reissue_pt_warnings(caught_warnings)
        elif self.deepspeed:
            pass  # deepspeed.save_checkpoint above saves model/optim/sched
        elif self.is_world_process_zero():
            torch.save(self.optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
            with warnings.catch_warnings(record=True) as caught_warnings:
                torch.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
            reissue_pt_warnings(caught_warnings)
```

2.
```
        if is_torch_tpu_available():
            xm.rendezvous("saving_optimizer_states")
            xm.save(self.optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
            with warnings.catch_warnings(record=True) as caught_warnings:
                xm.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
                reissue_pt_warnings(caught_warnings)
        elif self.is_world_process_zero() and not self.deepspeed:
            # deepspeed.save_checkpoint above saves model/optim/sched
            torch.save(self.optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
            with warnings.catch_warnings(record=True) as caught_warnings:
                torch.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
            reissue_pt_warnings(caught_warnings)
```

I'm pretty sure you'd prefer (2), correct?

Reply:
I have a small preference for (2), now that I understand it's a necessary branch. So if you have a strong preference for one, go ahead.

Reply:
Great. I have no problem with 2, since this is the main way of the project.

As you can tell my mind optimizes things, so when it's not a vertical branching it has to work much harder at understanding what's going on. Binary search in nested branches is a way faster for my brain to parse than sequential style `transformers` uses. and of course, new lines to separate different branch sets.

Reply:
Thank you for spotting that buglet, @sgugger